cobalt/updater/updater_module.cc

```diff
@@ -0,0 +1,560 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/updater/updater_module.h"
+
+#include <utility>
+#include <vector>
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+#include "base/files/file_path.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/functional/callback_helpers.h"
+#include "base/logging.h"
+#include "base/rand_util.h"
+#include "base/run_loop.h"
+#include "base/stl_util.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/task/sequenced_task_runner.h"
+#include "base/task/task_runner.h"
+#include "base/threading/scoped_blocking_call.h"
+#include "base/threading/thread.h"
+#include "base/time/time.h"
+#include "base/version.h"
+#include "cobalt/browser/switches.h"
+#include "cobalt/updater/util.h"
+#include "components/crx_file/crx_verifier.h"
+#include "components/update_client/cobalt_slot_management.h"
+#include "components/update_client/update_client_errors.h"
+#include "components/update_client/utils.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "starboard/common/file.h"
+#include "starboard/configuration_constants.h"
+#include "starboard/extension/installation_manager.h"
+
+namespace {
+
+using update_client::CobaltSlotManagement;
+using update_client::ComponentState;
+using update_client::UpdateCheckError;
+
+// The SHA256 hash of the "cobalt_evergreen_public" key.
+constexpr uint8_t kCobaltPublicKeyHash[] = {
+    0x51, 0xa8, 0xc0, 0x90, 0xf8, 0x1a, 0x14, 0xb0, 0xda, 0x7a, 0xfb,
+    0x9e, 0x8b, 0x2d, 0x22, 0x65, 0x19, 0xb1, 0xfa, 0xba, 0x02, 0x04,
+    0x3a, 0xb2, 0x7a, 0xf6, 0xfe, 0xd5, 0x35, 0xa1, 0x19, 0xd9};
+
+CobaltExtensionUpdaterNotificationState
+ComponentStateToCobaltExtensionUpdaterNotificationState(
+    ComponentState component_state) {
+  switch (component_state) {
+    case ComponentState::kChecking:
+      return kCobaltExtensionUpdaterNotificationStateChecking;
+    case ComponentState::kCanUpdate:
+      return kCobaltExtensionUpdaterNotificationStateUpdateAvailable;
+    case ComponentState::kDownloading:
+      return kCobaltExtensionUpdaterNotificationStateDownloading;
+    case ComponentState::kDownloaded:
+      return kCobaltExtensionUpdaterNotificationStateDownloaded;
+    case ComponentState::kUpdating:
+      return kCobaltExtensionUpdaterNotificationStateInstalling;
+    case ComponentState::kUpdated:
+      return kCobaltExtensionUpdaterNotificationStateUpdated;
+    case ComponentState::kUpToDate:
+      return kCobaltExtensionUpdaterNotificationStateUpToDate;
+    case ComponentState::kUpdateError:
+      return kCobaltExtensionUpdaterNotificationStateUpdateFailed;
+    default:
+      return kCobaltExtensionUpdaterNotificationStateNone;
+  }
+}
```

Comment:
nit: Don't have a default, and instead enumerate all ComponentStates?
```
    case ComponentState::kBla:
    case ComponentState::kFoo:
      return kCobaltExtensionUpdaterNotificationStateNone;
```
where kBla, kFoo are the ones not handled already. This makes
sure that if someone was to add a ComponentState in the future,
the code here would need to be updated.